dp_check/dp_check.go

```diff
@@ -142,31 +233,101 @@ func main() {
+		infoLog.Printf("Check if this VM enables DirectPath by sending http GET request to %s", metadataServerUrl)
+		req, err := http.NewRequest("GET", metadataServerUrl, nil)
+		if err != nil {
+			return err
+		}
+		req.Header.Add("Metadata-Flavor", "Google")
+		resp, err := client.Do(req)
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+		ipv6FromMetadataServer = net.ParseIP(strings.TrimSuffix(string(body), "\n"))
+		if err != nil {
+			return err
+		}
+		if resp.StatusCode == 200 {
+			return nil
+		}
+		if resp.StatusCode == 404 {
+			return fmt.Errorf("This VM doesn't have DirectPath access")
+		}
+		return fmt.Errorf("Received status code %d in response to metadata server GET request to URL: %s. This is unexpected (we only expect status codes 200 or 404), and so this may indicate a bug", resp.StatusCode, metadataServerUrl)
+	})
+
+	var directPathNetworkInterface net.Interface
 	runCheck("IPv6 addresses", func() error {
-		var out string
-		var err error
-		cmd := "ip -6 addr show"
-		infoLog.Printf("Check IPv6 addresses with:|%v|...\n", cmd)
-		if out, err = getOutput(cmd); err != nil {
+		ifaces, err := net.Interfaces()
+		if err != nil {
 			return err
 		}
-		if !strings.Contains(out, "inet6 2600") {
-			return fmt.Errorf("This VM is missing a global 2600-prefixed IPv6 address. IPv6 DHCP setup either failed or hasn't been attempted")
+		// Go through all interfaces on the VM to see if IPv6 is enabled and if there is an IPv6 address, then check against the IPv6 address returned from metadataserver
+		for _, iface := range ifaces {
+			if iface.Flags&net.FlagLoopback != 0 {
+				continue
+			}
+			if iface.Flags&net.FlagUp != net.FlagUp {
+				continue
+			}
+			infoLog.Printf("Checking non-loopback and up network interface: |Name: %s, hardware address: %s, flags: %s\n", iface.Name, iface.HardwareAddr, iface.Flags)
+			ifaddrs, err := iface.Addrs()
+			if err != nil {
+				return err
+			}
+			for _, ifaddr := range ifaddrs {
+				ip := ifaddr.(*net.IPNet).IP
+				if ip.To4() == nil && ip.Equal(ipv6FromMetadataServer) {
+					directPathNetworkInterface = iface
+					break
+				}
+			}
+			if directPathNetworkInterface.Name != "" {
+				break
+			}
+		}
+		// There is IPv6 address on the machine but it does not match what metadata server returns
+		if directPathNetworkInterface.Name == "" {
+			return fmt.Errorf("This VM was expected to have a network interface with IPv6 address: %s assigned to it, but no such interface was found, IPv6 DHCP setup either failed or hasn't been attempted", ipv6FromMetadataServer)
+		}
+		infoLog.Printf("Found the valid directpath network interface %s with hardware address |%s| and flag %s\n", directPathNetworkInterface.Name, directPathNetworkInterface.HardwareAddr, directPathNetworkInterface.Flags)
+		return nil
+	})
+
+	runCheck("IPv6 routes", func() error {
+		// Check if there is route to the gRPCLB
```

Comment:
nit: inaccuate comment, as this route is used for all backends rather than only for grpclb

Reply:
also, can we remove this comment and turn it into a log message that details implementation of <i>how</i> we check IPv6 routes ? (helps to give more context for debugging)